src/gt4py/next/common.py

```diff
@@ -84,31 +74,104 @@ def __str__(self):
         return f"{self.value}[{self.kind}]"
 
 
+class OpenBound(enum.Enum):
+    LOWER = enum.auto()
+    UPPER = enum.auto()
+
+    def __add__(self, _: int) -> Self:
+        return self
+
+    __radd__ = __add__
+
+    def __sub__(self, _: int) -> Self:
+        return self
+
+    __rsub__ = __sub__
+
+    def __le__(self, other: int | OpenBound) -> bool:
+        return self is self.LOWER or other is self.UPPER
+
+    def __lt__(self, other: int | OpenBound) -> bool:
+        return self is self.LOWER and other is not self
+
+    def __ge__(self, other: int | OpenBound) -> bool:
+        return self is self.UPPER or other is self.LOWER
+
+    def __gt__(self, other: int | OpenBound) -> bool:
+        return self is self.UPPER and other is not self
+
+
 @dataclasses.dataclass(frozen=True, init=False)
-class UnitRange(Sequence[int], Set[int]):
-    """Range from `start` to `stop` with step size one."""
+class UnitRange(Sequence[int]):
+    """
+    Range from `start` to `stop` with step size one.
 
-    start: int
-    stop: int
+    An open range is constructed by passing `None` for `start` and/or `stop`.
+    """
+
+    start: int | OpenBound
+    stop: int | OpenBound
 
-    def __init__(self, start: core_defs.IntegralScalar, stop: core_defs.IntegralScalar) -> None:
-        if start < stop:
-            object.__setattr__(self, "start", int(start))
-            object.__setattr__(self, "stop", int(stop))
+    def __init__(
+        self,
+        start: core_defs.IntegralScalar | OpenBound | None,
+        stop: core_defs.IntegralScalar | OpenBound | None,
+    ) -> None:
```

Comment:
Just a comment: it seems that `None` values are supported in the initializer args just as a convenient way for users to specify open ranges, but it makes the implementation slightly more complex than strictly needed. An alternative could be to create two builder methods for the cases where the lower or upper bounds are open, in the same style as the `unbounded()` constructor. For example:
```python
@classmethod
def stop_at(cls, stop: int) -> UnitRange:
    return UnitRange(OpenBound.LOWER, stop)

@classmethod
def start_at(cls, start int) -> UnitRange:
    return UnitRange(start, OpenBound.UPPER)

```

Reply:
After you suggested changes, it's still slightly more complex than needed, but also relatively straight-forward. Do you still think we need the builders?